kafka-connect-elastic/src/main/scala/com/datamountaineer/streamreactor/connect/elastic/ElasticJsonWriter.scala

```diff
@@ -81,7 +86,25 @@ class ElasticJsonWriter(client: ElasticClient, settings: ElasticSettings) extend
 
         val indexes = sinkRecords
                         .map(r => convert(r, fields, ignoreFields))
-                        .map(r => index into i.toLowerCase / i source r)
+                        .map({ r =>
+                          configMap(r.topic).getWriteMode match {
+                            case WriteModeEnum.INSERT => index into i.toLowerCase / i source r
+                            case WriteModeEnum.UPSERT =>
+                              // Build a Struct field extractor to get the value from the PK field
+                              val pkField = settings.pks(r.topic)
+                              // Extractor includes all since we already converted the records to have only needed fields
+                              val extractor = StructFieldsExtractor(includeAllFields = true, Map(pkField -> pkField))
+                              val fieldsAndValues = extractor.get(r.value.asInstanceOf[Struct]).toMap
+                              if(fieldsAndValues.contains(pkField)){
+                                val pkValue = fieldsAndValues(pkField).toString
+                                update id pkValue in i.toLowerCase / i docAsUpsert fieldsAndValues
+                              } else {
+                                logger.error(s"Empty ${pkField}")
+                                null
```

Comment:
yeah, I added that because we _didn't_ want the error to explode the sink but I realise it's not your way :) I can revert that and we will add an extra check on our server side to ensure the field exist.